Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt

```diff
@@ -0,0 +1,135 @@
+import de.rki.coronawarnapp.risk.RiskLevel.UNDETERMINED
+import de.rki.coronawarnapp.risk.RiskLevel.UNKNOWN_RISK_INITIAL
+import de.rki.coronawarnapp.risk.RiskLevel.UNKNOWN_RISK_OUTDATED_RESULTS
+import de.rki.coronawarnapp.risk.RiskLevel.UNKNOWN_RISK_OUTDATED_RESULTS_MANUAL
+import de.rki.coronawarnapp.storage.RiskLevelRepository
+import de.rki.coronawarnapp.task.Task
+import de.rki.coronawarnapp.task.TaskCancellationException
+import de.rki.coronawarnapp.task.TaskFactory
+import de.rki.coronawarnapp.task.common.DefaultProgress
+import de.rki.coronawarnapp.util.ConnectivityHelper
+import de.rki.coronawarnapp.util.ConnectivityHelper.isNetworkEnabled
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import org.joda.time.Duration
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Provider
+
+class RiskLevelTask @Inject constructor(
+    private val riskLevels: RiskLevels,
+    private val context: Context
+) : Task<DefaultProgress, RiskLevelTask.Result> {
+
+    private val internalProgress = ConflatedBroadcastChannel<DefaultProgress>()
+    override val progress: Flow<DefaultProgress> = internalProgress.asFlow()
+
+    private var isCanceled = false
+
+    override suspend fun run(arguments: Task.Arguments) = try {
+        Timber.d("Running with arguments=%s", arguments)
+        Result(
+            // If there is no connectivity the transaction will set the last calculated
+            // risk level
+            if (isNetworkEnabled(context)) {
+                with(riskLevels) {
+                    when {
+                        calculationNotPossibleBecauseTracingIsOff().also {
+                            checkCancel()
+                        } -> NO_CALCULATION_POSSIBLE_TRACING_OFF
+
+                        calculationNotPossibleBecauseNoKeys.also {
+                            checkCancel()
+                        } -> UNKNOWN_RISK_INITIAL
+
+                        calculationNotPossibleBecauseOfOutdatedResults.also {
+                            checkCancel()
+                        } -> if (backgroundJobsEnabled)
+                            UNKNOWN_RISK_OUTDATED_RESULTS
+                        else
+                            UNKNOWN_RISK_OUTDATED_RESULTS_MANUAL
+
+                        isIncreasedRisk().also {
+                            checkCancel()
+                        } -> INCREASED_RISK
+
+                        !isActiveTracingTimeAboveThreshold.also {
+                            checkCancel()
+                        } -> UNKNOWN_RISK_INITIAL
+
+                        else -> LOW_LEVEL_RISK
+                    }.also {
+                        checkCancel()
+                        updateRepository(it, System.currentTimeMillis())
+                    }
+                }
+            } else {
+                RiskLevelRepository.setLastCalculatedRiskLevelAsCurrent()
```

Comment:
Maybe early abort instead of the big if-else?

Reply:
OK.